src/serialize-spec.ts

```diff
@@ -184,55 +188,141 @@ function serializeNode(node: Node): any {
   }
 }
 
-function serializeProvableType(type: ProvableType<any>): Record<string, any> {
+type SerializedType =
+  | { _type: O1jsTypeName }
+  | { _type: 'Struct'; properties: SerializedNestedType }
+  | { _type: 'Constant'; size: number }
+  | { _type: 'Bytes'; size: number }
+  | { _type: 'Proof'; proof: Record<string, any> }
+  | { _type: 'String' };
+
+type SerializedNestedType =
+  | SerializedType
+  | { [key: string]: SerializedNestedType };
+
+function serializeProvableType(type: ProvableType<any>): SerializedType {
   if ('serialize' in type && typeof type.serialize === 'function') {
     return type.serialize();
   }
-  // TODO: handle case when type is a Struct
-  const typeName = mapProvableTypeToName.get(type);
-  if (typeName === undefined) {
-    throw Error(`serializeProvableType: Unsupported provable type: ${type}`);
+  if ((type as any).prototype instanceof Bytes.Base) {
+    return { _type: 'Bytes', size: (type as typeof Bytes.Base).size };
   }
-  return { type: typeName };
+  if ((type as any).prototype instanceof DynamicProof) {
+    let { publicInputType, publicOutputType, maxProofsVerified, featureFlags } =
+      type as typeof DynamicProof;
+    let proof = {
+      name: (type as typeof DynamicProof).name,
+      publicInput: serializeProvableType(publicInputType),
+      publicOutput: serializeProvableType(publicOutputType),
+      maxProofsVerified,
+      featureFlags: replaceUndefined(featureFlags),
+    };
+    return { _type: 'Proof', proof };
+  }
+  let _type = mapProvableTypeToName.get(type);
+  if (_type === undefined && (type as any)._isStruct) {
+    return serializeStruct(type as Struct<any>);
+  }
+  assert(
+    _type !== undefined,
+    `serializeProvableType: Unsupported provable type: ${type}`
+  );
+  return { _type };
```

Comment:
Changed the `type` field to `_type` for serialized provable types, so that it's distinguished from nested types which also happen to have a "type" field